lxd/device/gpu.go

```diff
@@ -0,0 +1,644 @@
+
+	allGpus := d.deviceWantsAllGPUs(d.config)
+	gpus, nvidiaDevices, err := d.deviceLoadGpu(allGpus)
+	if err != nil {
+		return nil, err
+	}
+
+	sawNvidia := false
+	found := false
+	for _, gpu := range gpus {
+		if (d.config["vendorid"] != "" && gpu.vendorID != d.config["vendorid"]) ||
+			(d.config["pci"] != "" && gpu.pci != d.config["pci"]) ||
+			(d.config["productid"] != "" && gpu.productID != d.config["productid"]) ||
+			(d.config["id"] != "" && gpu.id != d.config["id"]) {
+			continue
+		}
+
+		found = true
+		err := unixDeviceSetupCharNum(d.state, d.instance.DevicesPath(), "unix", d.name, d.config, gpu.major, gpu.minor, gpu.path, false, &runConf)
+		if err != nil {
+			return nil, err
+		}
+
+		if !gpu.isNvidia {
+			continue
+		}
+
+		if gpu.nvidia.path != "" {
+			err = unixDeviceSetupCharNum(d.state, d.instance.DevicesPath(), "unix", d.name, d.config, gpu.nvidia.major, gpu.nvidia.minor, gpu.nvidia.path, false, &runConf)
+			if err != nil {
+				return nil, err
+			}
+		} else if !allGpus {
+			return nil, fmt.Errorf("Failed to detect correct \"/dev/nvidia\" path")
+		}
+
+		sawNvidia = true
+	}
+
+	if sawNvidia {
+		for _, gpu := range nvidiaDevices {
+			instanceConfig := d.instance.ExpandedConfig()
+
+			// No need to mount additional nvidia non-card devices as the nvidia.runtime
+			// setting will do this for us.
+			if shared.IsTrue(instanceConfig["nvidia.runtime"]) {
+				if !gpu.isCard {
+					continue
+				}
+			}
+
+			prefix := unixDeviceJoinPath("unix", d.name)
+			if UnixDeviceExists(d.instance.DevicesPath(), prefix, gpu.path) {
+				continue
+			}
+
+			err = unixDeviceSetupCharNum(d.state, d.instance.DevicesPath(), "unix", d.name, d.config, gpu.major, gpu.minor, gpu.path, false, &runConf)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("Failed to detect requested GPU device")
+	}
+
+	return &runConf, nil
+}
+
+// Stop is run when the device is removed from the instance.
+func (d *gpu) Stop() (*RunConfig, error) {
+	runConf := RunConfig{
+		PostHooks: []func() error{d.postStop},
+	}
+
+	err := unixDeviceRemove(d.instance.DevicesPath(), "unix", d.name, &runConf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &runConf, nil
+}
+
+// postStop is run after the device is removed from the instance.
+func (d *gpu) postStop() error {
+	// Remove host files for this device.
+	err := unixDeviceDeleteFiles(d.state, d.instance.DevicesPath(), "unix", d.name)
+	if err != nil {
+		return fmt.Errorf("Failed to delete files for device '%s': %v", d.name, err)
+	}
+
+	return nil
+}
+
+// deviceWantsAllGPUs whether the LXD device wants to passthrough all GPUs on the host.
+func (d *gpu) deviceWantsAllGPUs(m map[string]string) bool {
+	return m["vendorid"] == "" && m["productid"] == "" && m["id"] == "" && m["pci"] == ""
+}
+
+// deviceLoadGpu probes the system for information about the available GPUs.
+func (d *gpu) deviceLoadGpu(all bool) ([]gpuDevice, []nvidiaGpuDevice, error) {
```

Comment:
Yeah, looks like we can likely get rid of this entire thing in favor of a call to `resources.GetGPU()`

Reply:
@stgraber I'll add a card for porting over this and infiniband to the resources package.